File: Misc/Dall-E/nodejs/dalle-generate-no-module.js
const fs = require('fs');
const commandLineArgs = require('command-line-args');

const optionDefinitions = [
    { name: 'prompt', alias: 'p', type: String, defaultValue: "an isometric view of a miniature city, tilt shift, bokeh, voxel, vray render, high detail" },
    { name: 'number', alias: 'n', type: Number, defaultValue: 1 },
    { name: 'size', alias: 's', type: Number, defaultValue: 256 },
];

const options = commandLineArgs(optionDefinitions);

const key = process.env.OPENAI_API_KEY;
const dalleEndpoint = 'https://api.openai.com/v1/images/generations';

const reqBody = {
    prompt: options.prompt,
    n: options.number,
    size: `${options.size}x${options.size}`,
    response_format: 'b64_json',
};

const reqParams = {
    method: 'POST',
    headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${key}`,
    },
    body: JSON.stringify(reqBody)
}

// Start request/s.
// You will need Node v18.X.X to use the `fetch` command in Node.js
fetch(dalleEndpoint, reqParams)
    .then(res => res.json())
    .then(saveImages);


function saveImages(jsonData)
{
    const now = Date.now();
    for (let i = 0; i < jsonData.data.length; i++) {
        const b64 = jsonData.data[i].b64_json;
        const buffer = Buffer.from(b64, "base64");
        const filename = `image_${now}_${i}.png`;
        console.log("Writing image " + filename);
        fs.writeFileSync(filename, buffer);
    }
}

File: Misc/Dall-E/python/dalle-generate-no-module.py
# This example is part of the `Fun with Dall-E` video tutorial series
# at ParametricCamp: https://www.youtube.com/playlist?list=PLx3k0RGeXZ_zs3az0Z2BnpTIPH2lxQfFX
import os
import requests
import json
import base64, time
import argparse

# Initiate the parser
parser = argparse.ArgumentParser()
parser.add_argument("-p", "--prompt", help="Text to image prompt:", default='an isometric view of a miniature city, tilt shift, bokeh, voxel, vray render, high detail')
parser.add_argument("-n", "--number", help="Number of images generated", default=1)
parser.add_argument("-s", "--size", help="Image size: 256, 512 or 1024", default=256)

# Read arguments from the command line
args = parser.parse_args()

api_key = os.getenv("OPENAI_API_KEY")
url = 'https://api.openai.com/v1/images/generations'

custom_headers = {
    "Content-Type": "application/json",
    "Authorization": "Bearer " + api_key,
}

reqBody = {
    "prompt": args.prompt,
    "n": int(args.number),
    "size": f'{args.size}x{args.size}',
    "response_format": "b64_json"
}

res = requests.post(url, 
    data=json.dumps(reqBody),
    headers=custom_headers,
)

# print(r)
# print(r.url)
# print(r.status_code)
print(res.text)
# print(r.content)

res_json = json.loads(res.text)
for i in range(0, len(res_json['data'])):
    img_file_name = f'image_{int(time.time())}_{i}.png'
    with open(img_file_name, 'wb') as f:
        f.write(base64.urlsafe_b64decode(res_json['data'][i]['b64_json']))